### ISSUE TITLE
Question about Gaussian degeneration and the role of base-to-finetune strategy in TokenGS training

### ISSUE THREAD

**Issue by Langwenchong** [inspatio/querysplat#6]
Dear authors,

Thank you very much for open-sourcing such an excellent work. I noticed that your project further improves and extends TokenGS, achieving significantly better reconstruction quality. As I understand, this improvement mainly comes from the dual-branch strategy proposed in your paper: first predicting the Gaussian attributes using VGGT features, and then further predicting the color attributes by incorporating appearance features.

After observing the reconstruction results shown on your project page, I noticed that your Gaussian point clouds are distributed very uniformly, and the reconstructed high-frequency details exhibit very sharp boundaries. I speculate that this is partly due to the stable initialization of Gaussian positions provided by the VGGT-predicted point clouds, which helps stabilize the early training stage. However, I would also like to ask whether this improvement is related to the base → finetune training strategy proposed in your paper.

I am currently also trying to train TokenGS, but I have observed a significant issue during training: the Gaussian point clouds tend to suffer from insufficient differentiation. Specifically, as training progresses, only a small number of Gaussians continue to provide meaningful rendering contributions, while a large number of Gaussians gradually degenerate into states with opacity close to zero and extremely small scales (as shown in the figure below). Eventually, the PSNR improves to around 20 but becomes difficult to further increase.

<img width="960" height="914" alt="Image" src="https://github.com/user-attachments/assets/d1a24fd6-bc2b-4019-8739-6e805d64efa9" />

My current understanding is that this degeneration may be caused by the fact that Gaussian positions are highly clustered at the beginning of training, making it difficult for different Gaussians to differentiate effectively. If some Gaussians fail to quickly learn positions and attributes that contribute sufficiently to rendering in the early stage, they may gradually become occluded by other Gaussians and receive fewer effective gradients during later optimization. As a result, they may eventually collapse into low-opacity and small-scale Gaussians.

Therefore, I would like to ask whether your base → finetune strategy is also designed to alleviate this issue. My understanding is that, in the base stage, you first train with a smaller number of GS tokens, and then during progressive finetuning, you increase the number of GS tokens by duplicating the already optimized GS tokens from the base model and adding small perturbations for initialization. Is this strategy intended to provide newly added Gaussians with better initialization based on a stable Gaussian distribution, thereby preventing a large number of Gaussians from failing to learn useful representations and degenerating during the early training stage?

Additionally, I would like to ask about the expected PSNR curve during normal GS training. For a 4-view input setting, should the PSNR curve generally continue increasing without obvious convergence before reaching around 22–23? Or is it also possible that reaching a plateau around 20 is a normal phenomenon?

Thank you very much for your time and for sharing your insights. I would greatly appreciate any advice or clarification you could provide.

Best regards.

**Comment by dragonylee** [comment]
Thanks for the detailed analysis!

For the Gaussian degeneration issue, the main thing that helped in our case is the early-stage regularization. Since this is a fully query-based formulation with a very high degree of freedom, the model can sometimes fit the rendering loss too early by simply reducing the opacity of poorly initialized Gaussians, instead of learning a meaningful Gaussian distribution. To avoid this, we add Chamfer distance and opacity regularization in the early stage of training. We also have fairly thorough ablations in the paper showing that these regularizations are important for stabilizing the Gaussian distribution.

For the base → finetune strategy, we didn't observe an obvious improvement in the Gaussian distribution itself. Our understanding is actually the other way around: the base model needs to learn a reasonably good Gaussian distribution first, so that the benefit of finetuning can be better realized.

For reference, in our 4-view base training, PSNR can usually reach around 23–24. But this curve depends on many factors, including the data sampling strategy, so this number should only be taken as a reference.

**Comment by Langwenchong** [comment]
Thank you very much for your reply. I would also like to ask a further question regarding the implementation details of the regularization loss. Do you apply the supervision to the entire point cloud, or do you use any sampling strategy during training?

Considering the computational cost, I currently randomly sample 5k points from both the predicted point cloud and the VGGT point cloud, and compute the Chamfer distance loss between these two subsets. I would really appreciate it if you could share some details about how you implement the regularization loss during training. Thank you very much!

**Comment by dragonylee** [comment]
In the base stage, we use 65,536 Gaussians, so we also sample 65,536 depth points from VGGT-Omega based on local patches.

We don't expect the final Gaussians to strictly align with the VGGT point cloud. So this supervision is only used in the early stage, while the final Gaussian distribution should still be learned from the rendering loss.

**Comment by Langwenchong** [comment]
Dear Author,

Sorry to bother you again, but I would like to ask a few more detailed questions about this issue.

After adding the regularization losses described in your paper, I found that after around 40K steps—when the regularization losses have been gradually annealed to zero—and after continuing training with the rendering loss for some time, I still observe that a certain number of Gaussians collapse or degenerate. This seems somewhat different from the result shown in Fig. 7 of your paper.

Below is the Gaussian distribution after applying the regularization losses, annealing them out, and then continuing training with the rendering loss for a while:

<img width="1132" height="775" alt="Image" src="https://github.com/user-attachments/assets/17d5d231-15d6-4f55-b327-9524dbadc643" />

<img width="510" height="433" alt="Image" src="https://github.com/user-attachments/assets/4f492912-2e2b-4063-9900-28812bfbcb92" />

As highlighted by the red boxes, there still seem to be many Gaussians clustered very densely in certain regions. However, according to Fig. 7, my understanding is that after applying the Early Reg strategy, this kind of severe Gaussian clustering or degeneration should be largely avoided. Therefore, I would like to ask whether the behavior I am observing is actually expected. If I continue training for a longer time, will these densely clustered Gaussians gradually spread out and become useful?

I also noticed that both TokenGS and your QuerySplat implementation apply an opacity-based filtering step when exporting the Gaussian point cloud for visualization. Is this mainly because the raw predicted Gaussian set still contains many low-opacity Gaussians, so that only Gaussians with meaningful rendering contributions are retained for visualization?

I further visualized the Gaussian-token clusters in my current scene, where Gaussians decoded from the same gs_token are assigned the same color. Interestingly, I found that Gaussians generated by different tokens often overlap with each other, and there does not seem to be very strong spatial locality for individual tokens.

<img width="1157" height="612" alt="Image" src="https://github.com/user-attachments/assets/ca329255-6a78-4c48-931b-c88523426396" />

This makes me wonder whether the remaining Gaussian collapse may partly come from the difficulty of requiring a single token to decode 64 different Gaussians. Even with the Early Reg losses, perhaps the main benefit is to help the Gaussian tokens differentiate more quickly during the early stage of training and learn the coarse scene structure, but it still cannot completely prevent some Gaussians decoded from the same token from collapsing or becoming redundant.

For reference, my implementation of the regularization losses is as follows:

<img width="693" height="809" alt="Image" src="https://github.com/user-attachments/assets/626da1d4-905d-4a9c-8cb4-74be78922a7c" />

I use a bidirectional Chamfer distance implemented with PyTorch3D.
For the opacity regularization, I enforce an initial gs_opacity_floor = 0.1.

After adding these two Early Reg terms, I indeed observe that the model learns the scene structure much faster than vanilla TokenGS, and a noticeably larger proportion of the predicted Gaussians actually contribute to rendering. However, my understanding is that Chamfer distance itself still cannot explicitly prevent multiple Gaussians from collapsing to nearby or identical positions. It mainly encourages the predicted point cloud to match the target geometry globally, so it may alleviate the problem but cannot fundamentally eliminate Gaussian clustering. Is this interpretation correct?

I also have another question regarding the later training stage. After the base training is completed, you increase the number of queries from 1024 to 8192, while each gs_token still decodes 64 Gaussians. My understanding is that this process is essentially initialized by duplicating or expanding the existing tokens, such that the newly introduced tokens start from representations similar to the original ones and are then further optimized to capture higher-frequency scene details.

If so, would it be reasonable to interpret the overall strategy in the following way: although it may still be difficult to completely avoid partial Gaussian collapse or redundancy, the Early Reg losses allow each token to decode a higher proportion of useful Gaussians during the base stage, and then increasing the number of tokens further increases the total number of effective Gaussians available for representing high-frequency details?

Finally, both TokenGS and QuerySplat seem to use roughly 1024 × 64 Gaussians, or even more after query expansion. However, I recently noticed the paper C3G: Learning Compact 3D Representations with 2K Gaussians, which reportedly achieves comparable PSNR using only around 2K Gaussians, or at most around 32K Gaussians depending on the setting. Does this suggest that current query-based methods such as TokenGS and QuerySplat still contain a considerable amount of Gaussian redundancy, and that only a subset of the predicted Gaussians are actually essential for rendering?

Sorry that my description is a bit long and perhaps somewhat scattered. I would really appreciate it if you could clarify whether my understanding above is reasonable, especially regarding the role of Early Reg, Gaussian redundancy/collapse, and the motivation behind the later query expansion.

Thank you very much for your time and help!

**Comment by dragonylee** [comment]
Thanks for the detailed follow-up 😄. A few clarifications from our side:

1. For the early regularization, in the base stage we decode 65,536 Gaussians and sample the same number of VGGT-Omega depth points using local patches for the bidirectional Chamfer loss. We haven't specifically studied what happens to the GS distribution after the regularization is annealed out. One possible explanation is that TokenGS learns its geometric features during training, so the early CD loss does not necessarily guarantee that the geometry learned later will remain consistent with the early-stage distribution. In our case, the geometry encoder is pretrained and fixed, which may make this prior more stable. This is just our speculation, though.

2. For opacity filtering, the main reason is cleaner visualization. We found that very low-opacity Gaussians can appear somewhat foggy in SuperSplat, so we filter them when exporting. Fig. 7 in the paper, however, uses all generated Gaussians without filtering. The inference script also provides an option `--save_gaussian_alpha_distribution` if you want to inspect this directly.

3. There is definitely some redundancy when one token decodes 64 Gaussians. We also tried 16 Gaussians per token; 64 still gave better metrics, although the gap was not very large, so we kept 64. Some overlap between Gaussians from different tokens also seems reasonable to us because enforcing too much spatial separation could potentially introduce gaps in the scene. For efficiency-oriented applications, using significantly fewer Gaussians with a small quality drop may indeed be a better trade-off, but that was not the main constraint we were optimizing for.

4. And yes, for the base → finetune strategy, an important prerequisite is that the base model is already trained as well as possible. The quality of the base representation largely determines how much benefit the later query expansion can provide.

**Comment by Langwenchong** [comment]
Thank you very much for your reply. I don’t have any further questions now. ^_^!
